### PR TITLE
Stage 3 summary

### PR DESCRIPTION
**Pull request by JacekGlen**

### CODE REVIEW

File: docs/report_stage_iii.md [hunk 1]
```diff
@@ -242,17 +253,25 @@ Sample results:
...
 
 As seen on the charts, the execution times for `ADD`, `DIV` and `MULMOD` are not consistent. While the majority behaves in a linear fashion, there is a large portion that often takes longer. Again, we attribute it to the JVM garbage collector or other JVM internals.
 
-The execution time for `EXP` increases linearly with the size of the second argument. Again the separation into two different lines shows that the execution time depends not only on the size of the argument but also its value. In contrast to other implementations, the second line has a constant slope.
+The execution time for `EXP` increases linearly with the size of the second argument. Again the separation into two different lines shows that the execution time depends not only on the size of the argument but also on its value. In contrast to other implementations, the second line has a constant slope.
+
+Some common OPCODEs perform worse than expected. Improving them would have a positive impact on the overall execution time. These are `LT`, `GT`, `AND`, `OR`, `XOR` and `BYTE`.
```
Comment (lukasz-glen) [inline docs/report_stage_iii.md:264]: I would avoid the terms "worse", "improve performance". Improving the execution time is not our goal. The fact that estimations for these OPCODEs are relatively higher can mean that there is a space for optimizations or other OPCODEs are very efficiently implemented.

**Review by lukasz-glen — COMMENTED** [review]
The first part

File: docs/report_stage_iii.md [hunk 2]
```diff
@@ -280,21 +299,27 @@ Sample results:
...
+
+Investigate the performance of the `BYTE`, `SHL`, `SHR` and `SAR` OPCODEs.
 
-Full etails:
-- [Analysis](./report_stage_iii_assets/final_estimation.html)
+## Measurement Analysis
+
+Full details:
+- [Gas cost estimation analysis](./report_stage_iii_assets/final_estimation.html)
 - [Alternative gas cost schedule](./report_stage_iii_assets/gas_schedule_comparison.csv)
```
Comment (lukasz-glen) [inline docs/report_stage_iii.md:322]: Is it scaled? Just to be sure.

File: docs/report_stage_iii.md [hunk 3]
```diff
@@ -280,21 +299,27 @@ Sample results:
...
 
 Rust EVM results are very consistent and follow the expected pattern. The engine overhead is rather small, which is expected from a Rust implementation.
 
-In some opcodes like `ADD`, `DIV` and `MULMOD` there is a visible pattern of the first execution being slower than the rest. After running the specific opcode for a few times, the execution time stabilizes. This might be due to some caching or other Rust internals.
+In some opcodes like `ADD`, `DIV` and `MULMOD` there is a visible pattern of the first execution being slower than the rest. After running the specific opcode a few times, the execution time stabilizes. This might be due to some caching or other Rust internals.
 
 This implementation does not have a separation of lines seen for the EXP arguments in other implementations.
 
-## Conclusions
+Some common OPCODEs perform worse than expected. Improving them would have a positive impact on the overall execution time. These are `BYTE`, `SHL`, `SHR` and `SAR`.
```
Comment (lukasz-glen) [inline docs/report_stage_iii.md:312]: The same. I would avoid "worse" and expressing a need for an improvement.

File: docs/report_stage_iii.md [hunk 4]
```diff
@@ -309,144 +334,257 @@ Calculating the averages, we get the following alternative gas cost schedule:
...
-SWAP16|3|1.63|-45.50%|19.30%
-
-Notes:
-
-[*] - Alghouth the data suggest a change, but these particular OPCODEs were not supposed to generate such high costs, especially `SELFBALANCE`. More investigation needed.
-
-[**] - With the introduction of new relative jumps and disallowing dynamic jumps (EIP-4750 and EIP-4200) these OPCODEs might become irrelevant.
-
-> **Final Remark**
+The raw results of the alternative gas cost schedule have been re-scaled to fit the current gas cost schedule using manually selected OPCODEs as the base. Our analysis showed that the scale of 1.4 gives the best fit.
```
Comment (lukasz-glen) [inline docs/report_stage_iii.md:339]: One sentence why we think 1.4 is a right choice.

File: docs/report_stage_iii.md [hunk 5]
```diff
@@ -309,144 +334,257 @@ Calculating the averages, we get the following alternative gas cost schedule:
...
+
+> **Remarks**
 >
 > The data clearly shows that some changes can be made to the current gas proposal. Still, any such changes should be carefully considered, as there are other factors not taken into account in this research: ease of implementation, backward compatibility, hard fork requirement, existing tooling, hardcoded values in auxiliary software, etc.
 >
 >The alternative gas cost schedule is a proposal and should be treated as such. We believe that the methodology presented in this report is a good starting point for further research and discussion.
+
+
+### EXP
+For the detailed analysis of the `EXP` opcode, we have used the following data:
```
Comment (lukasz-glen) [inline docs/report_stage_iii.md:480]:
```suggestion
Notice that `EXP` opcode execution cost is expressed with the formula: the base cost (currently 10) and a dynamic cost for each byte of an exponent (currently 50 for each byte). For the detailed analysis of the `EXP` opcode, we have used the following data:
```

File: docs/report_stage_iii.md [hunk 6]
```diff
@@ -309,144 +334,257 @@ Calculating the averages, we get the following alternative gas cost schedule:
...
+- [Geth](./report_stage_iii_assets/geth_measure_arguments_single_exp.html)
+- [Netehrmind](./report_stage_iii_assets/nethermind_measure_arguments_single_exp.html)
+- [EtherumJS](./report_stage_iii_assets/ethereumjs_measure_arguments_single_exp.html)
+- [Erigon](./report_stage_iii_assets/erigon_measure_arguments_single_exp.html)
+- [Besu](./report_stage_iii_assets/besu_measure_arguments_single_exp.html)
+- [Rust EVM](./report_stage_iii_assets/revm_measure_arguments_single_exp.html)
+
+Please see 'EXP' section in the corresponding files.
+
+The `EXP` opcode proved quite difficult to estimate due to high client variability. Additionally, the measure marginal method was not able to sufficiently estimate the base cost of the opcode. The `EXP` measure marginal method used the argument size of 1 byte, as this provided the most consistent results. Still, for some clients, the measured cost was below 0.
```
Comment (lukasz-glen) [inline docs/report_stage_iii.md:491]: "Still, for some clients, the measured cost was below 0." This the outcome of the measure arguments method, not the measure marginal method as suggested.

File: docs/report_stage_iii.md [hunk 7]
```diff
@@ -309,144 +334,257 @@ Calculating the averages, we get the following alternative gas cost schedule:
...
+
+For a more accurate estimation, we compare EXP base and dynamic times to the execution of `ADD` and `MUL` opcodes. These opcodes are used as a reference as they are the most basic arithmetic ones. Given this data we can estimate the cost of the single byte exponentiation as follows:
```
Comment (lukasz-glen) [inline docs/report_stage_iii.md:493]: Not "times", they are estimates.

File: docs/report_stage_iii.md [hunk 8]
```diff
@@ -309,144 +334,257 @@ Calculating the averages, we get the following alternative gas cost schedule:
...
+:----- | ----: | -----: | ----: | -----: | -----: | -----:
+EvmOne|18.30|30.34|214.25|35.23|15.35|0
+Geth|19.78|28.21|98.74|16.46|82.88|0
+Netehrmind|31.44|61.78|274.48|23.55|870.21|51.12
+EtherumJS|290|381|1590|18.96|4623|36.16
+Erigon|6.68|7.81|84.63|46.72|0|0
+Besu*|58815|75348|154635|9.22|415094|15.53
+Rust EVM|2.86|3.59|35.13|43.57|15.38|0
+
+*Besu results inflated by the internal loop of 10k times.
```
Comment (lukasz-glen) [inline docs/report_stage_iii.md:513]: Even more. Recall that results should not be compared between EVMs due to various benchmark methods used.

**Review by lukasz-glen — COMMENTED** [review]
The second part.
Should we make a comment on the missing PUSH0?

**Review by lukasz-glen — APPROVED** [review]

File: docs/report_stage_iii.md (hunk 2 above)
Comment (JacekGlen) [reply docs/report_stage_iii.md]: Renamed to reflect the actual data

**Pull request merged**